### PR TITLE
optimize workspace attribute + entity sql; support selected attributes [PERF-238; risk: low]

### PR DESCRIPTION
**Pull request by davidangb**
See also #1132 

This PR:
* adds a `fields` parameter to the list-workspace API, which allows the caller to specify which parts of the response should be returned
* skips the SQL query that powers workspace submission stats, if the user omitted them from the response
* skips the SQL query that powers workspace attributes, if the user omitted them from the response
* rewrites the workspace attributes SQL query away from Slick and into plain SQL, avoiding subqueries and full table scans
* rewrites the workspace attributes SQL query to request only specific attributes, if the user specified them

see the "Rawls - List Workspaces" section in https://docs.google.com/document/d/1JtDVZNRbqAjOCo1EunUBhdoY_pIB-lE3RfBMotars7I/edit#heading=h.pw1zyd3b01ec for fuller context.

-----

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 1]
```diff
@@ -57,6 +58,77 @@ trait WorkspaceComponent {
...
+      } else {
+        // we have workspaceIds, and we will be retrieving attributes.
+        val workspaceIdList = reduceSqlActionsWithDelim(workspaceIds.map { id => sql"$id" })
+
+        // does this need "AND WORKSPACE_ATTRIBUTE.deleted = FALSE"?
+        // the method this replaced did not specify that, so we also don't specify it here.
+        val startSql =
+          sql"""select
+            a.id, a.owner_id, a.namespace, a.name, a.value_string, a.value_number, a.value_boolean, a.value_json, a.value_entity_ref, a.list_index, a.list_length, a.deleted, a.deleted_date,
+            e_ref.id, e_ref.name, e_ref.entity_type, e_ref.workspace_id, e_ref.record_version, e_ref.deleted, e_ref.deleted_date
```
Comment (devonbush) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:100]: I suppose you could detect whether the join is needed for a performance improvement, but that probably won't buy you much given that it's (let's hope) indexed?  And you probably almost always need to filter out deleted ones?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiListOptionsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        assertResult(Set(
+          WorkspaceListResponse(WorkspaceAccessLevels.Owner, WorkspaceDetails(testWorkspaces.workspace.copy(lastModified = dateTime), Set.empty), Option(WorkspaceSubmissionStats(Option(testDate), Option(testDate), 2)), false),
+          WorkspaceListResponse(WorkspaceAccessLevels.Owner, WorkspaceDetails(testWorkspaces.workspace2.copy(lastModified = dateTime), Set.empty), Option(WorkspaceSubmissionStats(None, None, 0)), false)
+        )) {
+          responseAs[Array[WorkspaceListResponse]].toSet[WorkspaceListResponse].map(wslr => wslr.copy(workspace = wslr.workspace.copy(lastModified = dateTime)))
+        }
+      }
+  }
+
+  it should "return full response if querystring exists but no fields param" in withTestWorkspacesApiServices { services =>
```
Comment (devonbush) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiListOptionsSpec.scala:161]: this is some nice edge-case testing

**Review by devonbush — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 1 above)
Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:100]: I think the only way to detect if the join is needed would be to turn this into two queries, which I don't think is a win.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 3]
```diff
@@ -57,6 +58,77 @@ trait WorkspaceComponent {
...
+    def workspaceAttributesWithReferences(workspaceIds: Seq[UUID], attributeSpecs: WorkspaceAttributeSpecs): ReadAction[Seq[( (UUID, WorkspaceAttributeRecord), Option[EntityRecord] )]] = {
+      // if workspaceIds is empty, or the caller explicitly specified no attributes, short-circuit and return nothing
+      if (workspaceIds.isEmpty || (!attributeSpecs.all && attributeSpecs.attrsToSelect.isEmpty)) {
+        DBIO.successful(Seq())
+      } else {
+        // we have workspaceIds, and we will be retrieving attributes.
+        val workspaceIdList = reduceSqlActionsWithDelim(workspaceIds.map { id => sql"$id" })
+
+        // does this need "AND WORKSPACE_ATTRIBUTE.deleted = FALSE"?
+        // the method this replaced did not specify that, so we also don't specify it here.
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:96]: this is surprising. i'm trying to remember if this was intentional or not.

regardless, since you've left it equivalent to the current implementation, i don't see the problem with leaving it this way.

**Review by MatthewBemis — APPROVED** [review]
Looks good!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 3 above)
Comment (MatthewBemis) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:96]: (i'll let you know if I can remember the answer to this question and we can file a ticket if need be)
